### DISCUSSION TITLE
Everyday Usage: self-contained installation directory?

### DISCUSSION THREAD

**Discussion by WuMingIT** [Homebrew/discussions#4330] [Everyday usage]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 3.6.14
ORIGIN: https://github.com/Homebrew/brew
HEAD: 001bacee18c512657e1535dff757c8cfdaefc116
Last commit: 3 months ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 94be72df2da895478252fae6ca758dacfa2b32ac
Core tap last commit: 3 months ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 4
Homebrew Ruby: 2.6.8 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8_1/bin/ruby
CPU: quad-core 64-bit haswell
Clang: 13.0.0 build 1300
Git: 2.32.0 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 11.7.4-x86_64
CLT: 13.2.0.0.1.1638488800
Xcode: N/A
```

### Output of `brew doctor`

```shell
Warning: Broken symlinks were found. Remove them with `brew cleanup`:
  /usr/local/lib/docker/cli-plugins
```

### Description of issue

Firs of all thanks for the wonderful tool I have been using for many years.

I wish to keep my dev environment tidy. As such would like to install each package and its dependencies in self-contained directories. On another laptop lost track of what installed what and now is a tangled mess. 

Is it possible to define a custom installation directory and install each package and its dependencies there?

I am aware of the option to cleanup dependencies: https://stackoverflow.com/a/7333289
I am aware of the option to install `brew` itself anywhere: https://github.com/Homebrew/brew/blob/664d0c67d5947605c914c4c56ebcfaa80cb6eca0/docs/Installation.md#untar-anywhere

**Comment by SMillerDev** [comment]
> I wish to keep my dev environment tidy. As such would like to install each package and its dependencies in self-contained directories. On another laptop lost track of what installed what and now is a tangled mess.

All packages are in their own self-contained directories in `/usr/local/Cellar`. What you're suggesting will make you end up with a ton of installs of the same dependencies, at which point you have essentially reinvented Docker without the sandboxing.

> Is it possible to define a custom installation directory and install each package and its dependencies there?

It is possible to install brew multiple times in multiple locations, although unsupported: https://docs.brew.sh/Installation#untar-anywhere-unsupported

**Reply by WuMingIT** [reply]
> What you're suggesting will make you end up with a ton of installs of the same dependencies

Thanks for your reply. I understand this is contrary to the core concept of shared libraries or dependencies. But would like to have the choice to trade shared storage space with activity isolation. I am not completely aware of the tech ramifications: sw development is a hobby for me. But if VMs, Docker, SaaS and many other solutions are used by developers is also, my humble opinion, because of the want of keeping activities separated from each other. At the cost of data replication. Local or remote.

> at which point you have essentially reinvented Docker without the sandboxing

Funny. My question arises exactly because don't want to use Docker. With his layered fs on top of a Linux VM on top of APFS on top of macOS. Separate directories are fine with me.

**Reply by SMillerDev** [reply]
> But would like to have the choice to trade shared storage space with activity isolation

Putting software in different folders likely won't actually separate them. They will still provide the public interfaces that other programs can connect to through sockets, ports or your `$PATH` variable.

> With his layered fs on top of a Linux VM on top of APFS on top of macOS. 

FTR, it would actually just be a linux FS on top of the docker layer on top of APFS.

**Comment by WuMingIT** [comment]
> provide the public interfaces that other programs can connect to through sockets, ports or your $PATH variable.

What I mean and want is separate my activities in to different folders. So if I have to install a version of Ruby, of Cairo and of something else - for different experiments - they don't end up in the same `Cellar` directory. I am aware they can still communicate in a number of ways. Ideally I would redirect all attempts to `/dev/null` but that would be a tad excessive.

> on top of the docker layer

Docker on macOS includes a Linux VM. Mini or otherwise. But this is not the point so won't insist on this subject. Also am not an expert so this is also as far as I can go.

**Reply by SMillerDev** [reply]
> So if I have to install a version of Ruby, of Cairo and of something else - for different experiments - they don't end up in the same Cellar directory. 

Cairo I'm not sure about, but for Ruby people generally use RBEnv for this purpose.
